File: nodeServer/node/arenaapi.js
var https = require('https');
var util = require('util');
var url = require('url');

exports.url = '';
exports.log = true;                            // no console log by default
exports.colorOutput = process.stdout.isTTY;     // color if not directed to file
exports.sessionID = null;                       // set at login, used for others, back to null at logut

var showResult = function(info, statusCode, errors, result) {
  var inspectOptions = {depth: 4, colors: exports.colorOutput};
  if (errors != null) {
    var ansiRed  = exports.colorOutput ? ['\x1B[31m', '\x1B[39m'] : ['', ''];
    console.log('->%s %d %s ERRORS%s\n%s', ansiRed[0], statusCode, info, ansiRed[1], util.inspect(errors, inspectOptions));
  } else {
    var ansiBlue = exports.colorOutput ? ['\x1B[34m', '\x1B[39m'] : ['', ''];
    console.log('->%s %d %s RESULT%s\n%s', ansiBlue[0], statusCode, info, ansiBlue[1], util.inspect(result, inspectOptions));
  }
};

// apiCall is a function which takes the specifics of the particular API, builds the REST call, executes it, and calls the callback when complete
// name:     string api name for logging
// method:   string HTTP method: GET, POST, ...
// urlPart:  string Trailing part of REST url, depends on API call, <ARG> will be replaced by args.ARG
// args:     map    Arguments as nam / value pairs which will be inserted into the URL as part of the path or query arguments (may be null)
// body:     Buffer, string, or map Used for POST calls only as the POST data (may be null)
// callback: function(statusCode, errors, result) called when the API call is complete
//             statusCode is the integer HTTP status code
//             errors is the map (JSON object) returned from the API call
//             result is the map (JSON object) or Buffer returned from the API call

var apiCall = function(name, method, urlPart, args, body, callback) {
  var headers = {};
  if (args != null) {
    var query = [];     // these are the query parameters
    for (var key in args) {
      var k = urlPart.indexOf('<' + key + '>');
      if (k != -1)
        urlPart = urlPart.substring(0, k) + encodeURIComponent(args[key]) + urlPart.substring(k + 2 + key.length);
      else
        query.push(key + '=' + encodeURIComponent(args[key]));
    }
    if (query.length > 0)
      urlPart += '?' + query.join('&');
  }
  var bodyBuffer = null;        // REST body for POST calls
  if (body != null) {
    if (Buffer.isBuffer(body)) {
      bodyBuffer = body;
      headers['Content-Type'] = 'application/octet-stream';
    } else if (typeof body == 'string') {
      bodyBuffer = new Buffer(body, 'utf8');
      headers['Content-Type'] = 'text/plain';
    } else {
      bodyBuffer = new Buffer(JSON.stringify(body), 'utf8');
      headers['Content-Type'] = 'application/json';
    }
    headers['Content-Length'] = bodyBuffer.length;
  }

  if (exports.sessionID != null)
    headers['Cookie'] = 'arena_session_id=' + exports.sessionID;        //needed for all but a login

  var urlObj = url.parse(exports.url);

  var options = {
    hostname: urlObj.hostname,
    port: urlObj.port,
    path: urlObj.path + urlPart,
    method: method,
    headers: headers,
    rejectUnauthorized: false   // no SSL cert required
  };

  var req = https.request(options, function(res) {
    var buffers = [];
    res.on('data', function(data) { buffers.push(data); });
    res.on('end', function() {
      var buffer = buffers.length == 0 ? null : Buffer.concat(buffers);
      var bufferJSON = null;
      if (buffer != null && res.headers['content-type'].indexOf('application/json') != -1) { // node lowercases headers
        try {
          bufferJSON = JSON.parse(buffer.toString());
        } catch(e) {}
      }
      var errors = null;
      var result = null;
      if (res.statusCode >= 200 && res.statusCode < 300) {
        result = bufferJSON != null ? bufferJSON : buffer;
        if ('set-cookie' in res.headers && result != null && typeof result == 'object' && 'arenaSessionId' in result)        // is a login
          exports.sessionID = result.arenaSessionId;
      } else {
        errors = bufferJSON != null ? bufferJSON : {status: res.statusCode, errors: [{code: 9999, message: 'API RESPONSE ERROR', info: {method: method, url: options.path, args: args, headers: res.headers, result: buffer != null ? buffer.toString() : null}}]}; // format like API error
      }
      if (exports.log)
        showResult(options.method + ' ' + options.path + ' ' + name, res.statusCode, errors, result);
      if (errors == null && options.path.indexOf('/logout') != -1)
        exports.sessionID = null;       // export so api callers can test if session is valid (may have expired though)
      callback(res.statusCode, errors, result);
    });
  });

  req.on('error', function(e) {
    var errors = {status: 0, errors: [{code: 9998, message: 'API REQUEST ERROR: ' + e.message, info: {method: method, url: options.path, args: args}}]}; // format like API error
    if (exports.log)
      showResult(options.method + ' ' + options.path + ' ' + name, 0, errors, null);
    callback(0, errors, null);
  });

  if (bodyBuffer != null)
    req.write(bodyBuffer);
  req.end();

  if (exports.log)
    console.log('--------------------------------------------\n%s %s\nheaders: %j\nbody: %s', options.method, options.path, options.headers, bodyBuffer);
};

exports.apis = [
  {name: 'login',                 method: 'POST',   urlPart: 'login'},
  {name: 'logout',                method: 'GET',    urlPart: 'logout'},
  {name: 'getItemNumberFormats',  method: 'GET',    urlPart: 'item/numberformats'},
  {name: 'getItemCategories',     method: 'GET',    urlPart: 'item/categories'},
  {name: 'getFileCategories',     method: 'GET',    urlPart: 'file/categories'},
  {name: 'getItemAttributes',     method: 'GET',    urlPart: 'item/attributes'},
  {name: 'getItemNumberFormat',   method: 'GET',    urlPart: 'item/numberformats/<guid>'},
  {name: 'getItemRevisions',      method: 'GET',    urlPart: 'items/<guid>/revisions'},
  {name: 'getItemRequirements',   method: 'GET',    urlPart: 'items/<guid>/requirements'},
  {name: 'getItemRelationships',  method: 'GET',    urlPart: 'items/<guid>/relationships'},
  {name: 'getCategoryAttributes', method: 'GET',    urlPart: 'item/categories/<guid>/attributes'},
  {name: 'getItems',              method: 'GET',    urlPart: 'items'},
  {name: 'getItem',               method: 'GET',    urlPart: 'items/<guid>'},
  {name: 'updateItem',            method: 'PUT',    urlPart: 'items/<guid>'},
  {name: 'createItem',            method: 'POST',   urlPart: 'items'},
  {name: 'deleteItem',            method: 'DELETE', urlPart: 'items/<guid>'},
  {name: 'getItemBOM',            method: 'GET',    urlPart: 'items/<guid>/bom'},
  {name: 'getItemFiles',          method: 'GET',    urlPart: 'items/<guid>/files'},
  {name: 'getItemFileContent',    method: 'GET',    urlPart: 'items/<guid>/files/<fileguid>/content'},
  {name: 'addItemFile',           method: 'POST',   urlPart: 'items/<guid>/files'},
  {name: 'addItemFileContent',    method: 'POST',   urlPart: 'items/<guid>/files/<fileguid>/content'},
  {name: 'deleteItemFile',        method: 'DELETE', urlPart: 'items/<guid>/files/<fileguid>'}
];

exports.apis.forEach(function(api) {
  if (api.method == 'POST' || api.method == 'PUT')
    exports[api.name] = function(args, body, callback) { apiCall(api.name, api.method, api.urlPart, args, body, callback); };
  else
    exports[api.name] = function(args, callback) { apiCall(api.name, api.method, api.urlPart, args, null, callback); };
});

File: nodeServer/java/arenaapi/ArenaAPI.java
package arenaapi;

import java.io.*;
import java.util.*;
import java.net.*;
import java.security.*;
import java.security.cert.*;
import javax.net.ssl.*;

import json.*;

public class ArenaAPI {

    public String url = "";
    public boolean log = false;
    public String sessionID = null;

    static Map<String, String[]> apiMap = new HashMap<String, String[]>();

    static {
        apiMap.put("login", new String[]{"POST", "login"});
        apiMap.put("logout", new String[]{"GET", "logout"});
        apiMap.put("getItemNumberFormats", new String[]{"GET", "item/numberformats"});
        apiMap.put("getItemCategories", new String[]{"GET", "item/categories"});
        apiMap.put("getFileCategories", new String[]{"GET", "file/categories"});
        apiMap.put("getItemAttributes", new String[]{"GET", "item/attributes"});
        apiMap.put("getItemNumberFormat", new String[]{"GET", "item/numberformats/<guid>"});
        apiMap.put("getItemRevisions", new String[]{"GET", "items/<guid>/revisions"});
        apiMap.put("getItemRequirements", new String[]{"GET", "items/<guid>/requirements"});
        apiMap.put("getItemRelationships", new String[]{"GET", "items/<guid>/relationships"});
        apiMap.put("getCategoryAttributes", new String[]{"GET", "item/categories/<guid>/attributes"});
        apiMap.put("getItems", new String[]{"GET", "items"});
        apiMap.put("getItem", new String[]{"GET", "items/<guid>"});
        apiMap.put("updateItem", new String[]{"PUT", "items/<guid>"});
        apiMap.put("createItem", new String[]{"POST", "items"});
        apiMap.put("deleteItem", new String[]{"DELETE", "items/<guid>"});
        apiMap.put("getItemBOM", new String[]{"GET", "items/<guid>/bom"});
        apiMap.put("getItemFiles", new String[]{"GET", "items/<guid>/files"});
        apiMap.put("getItemFileContent", new String[]{"GET", "items/<guid>/files/<fileguid>/content"});
        apiMap.put("addItemFile", new String[]{"POST", "items/<guid>/files"});
        apiMap.put("addItemFileContent", new String[]{"POST", "items/<guid>/files/<fileguid>/content"});
        apiMap.put("deleteItemFile", new String[]{"DELETE", "items/<guid>/files/<fileguid>"});
    }

    public ArenaAPI() {
        disableCertificateValidation();
    }

    public Object apiCall(String name, JSONObject args, Object body) throws ArenaAPIException, IOException, URISyntaxException {
        HttpURLConnection connection = null;
        try {
            String[] apiInfo = apiMap.get(name);
            String method = apiInfo[0];
            String urlPart = apiInfo[1];
            String query = "";
            if (args != null) {
                char joinChar = '?';
                for (Map.Entry<String, JSONValue> arg : args.getValue().entrySet()) {
                    String key = arg.getKey();
                    String value = arg.getValue().stringValue();
                    int k = urlPart.indexOf("<" + key + ">");
                    if (k != -1) {
                        urlPart = urlPart.substring(0, k) + value + urlPart.substring(k + key.length() + 2);
                    } else {
                        query += joinChar + key + '=' + value;
                        joinChar = '&';
                    }
                }
            }

            Map<String, List<String>> requestProperties = null;

            URL u = new URL(url + urlPart);   // maybe need to escape query part
            URL urlObj = new URI(u.getProtocol(), u.getAuthority(), u.getPath(), query, null).toURL();

//      URL urlObj = new URL(url + urlPart + query);
            connection = (HttpURLConnection) urlObj.openConnection();
            connection.setRequestMethod(method);

            if (sessionID != null)
                connection.addRequestProperty("Cookie", "arena_session_id=" + sessionID);

            if (body != null) {
                byte[] bodyBuffer = body instanceof byte[] ? (byte[]) body : body.toString().getBytes("UTF-8");
                if (bodyBuffer.length > 0) {
                    connection.addRequestProperty("Content-Length", Integer.toString(bodyBuffer.length));
                    connection.addRequestProperty("Content-Type", body instanceof byte[] ? "application/octet-stream" : "application/json");
                    if (log)
                        requestProperties = connection.getRequestProperties();
                    connection.setDoOutput(true);
                    OutputStream os = connection.getOutputStream();
                    os.write(bodyBuffer);
                    os.close();
                }
            }

            if (log) {
                if (requestProperties == null)
                    requestProperties = connection.getRequestProperties();
                String b = body == null ? "null" : body instanceof byte[] ? "<byte[" + ((byte[]) body).length + "]>" : body.toString();
                System.out.println("--------------------------------------------\n" + method + " " + urlObj.getPath() + query + "\nheaders: " + requestProperties.toString() + "\nbody: " + b);
            }

            int statusCode = connection.getResponseCode();

            List<String> cookies = connection.getHeaderFields().get("Set-Cookie");
            if (cookies != null) {
                for (String cookie : cookies) {
                    int k = cookie.indexOf("arena_session_id=");
                    if (k != -1)
                        sessionID = URLDecoder.decode(cookie.substring(k + 17, cookie.indexOf(';')), "UTF-8");
                }
            }

            InputStream is = connection.getErrorStream();
            if (is == null)
                is = connection.getInputStream();
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            byte[] buffer = new byte[2048];
            int length;
            while ((length = is.read(buffer)) != -1)
                baos.write(buffer, 0, length);
            is.close();
            byte[] resultBytes = baos.toByteArray();

            JSONValue resultJSON = null;
            String contentType = connection.getHeaderField("Content-Type");
            if (contentType != null && contentType.toUpperCase().indexOf("JSON") != -1)
                resultJSON = JSONValue.parseJSONValue(new String(resultBytes, "UTF-8"));

            Object result = null;
            JSONObject errors = null;
            if (statusCode >= 200 && statusCode < 300) {
                result = resultJSON != null ? resultJSON : resultBytes;
            } else {
                if (resultJSON instanceof JSONObject)
                    errors = (JSONObject) resultJSON;
                else
                    errors = buildError(statusCode, 9999, "API RESPONSE ERROR: " + name + " statusCode: " + statusCode);
            }

            if (log)
                showResult(method + ' ' + urlObj.getPath() + query + ' ' + name, statusCode, errors, result);

            if (errors != null)
                throw new ArenaAPIException(statusCode, errors);

            return result;

        } catch (UnsupportedEncodingException e) {
            return null;
        } finally {
            if (connection != null)
                connection.disconnect();
        }
    }

    JSONObject buildError(int statusCode, int code, String message) {
        JSONObject errors = new JSONObject();
        errors.put("status", new JSONNumber(statusCode));
        JSONObject info = new JSONObject();
        info.put("code", new JSONNumber(code));
        info.put("message", new JSONString(message));
        JSONArray ja = new JSONArray(1);
        ja.add(info);
        errors.put("errors", ja);
        return errors;
    }

    void disableCertificateValidation() {
        try {
            HostnameVerifier hostnameVerifier = new HostnameVerifier() {
                public boolean verify(String urlHostName, SSLSession session) {
                    return true;
                }
            };
            HttpsURLConnection.setDefaultHostnameVerifier(hostnameVerifier);
            TrustManager[] trustManagers = new TrustManager[]{new X509TrustManager() {
                public X509Certificate[] getAcceptedIssuers() {
                    return null;
                }

                public void checkClientTrusted(X509Certificate[] certs, String authType) {
                }

                public void checkServerTrusted(X509Certificate[] certs, String authType) {
                }
            }};
            SSLContext sc = SSLContext.getInstance("SSL");
            sc.init(null, trustManagers, new SecureRandom());
            HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
        } catch (Exception e) {
        }
    }

    public void showResult(String info, int statusCode, JSONObject errors, Object result) {
        if (errors != null) {
            System.out.println("-> " + statusCode + ' ' + info + " ERRORS " + errors.toStringIndent("  "));
        } else {
            String s = result instanceof JSONObject ? ((JSONObject) result).toStringIndent("  ") : "<byte[" + ((byte[]) result).length + "]>";
            System.out.println("-> " + statusCode + ' ' + info + " RESULT " + s);
        }
    }

}
